admin/server/auth/claims.go

```diff
@@ -167,44 +192,64 @@ func (c *authTokenClaims) Can(ctx context.Context, orgID string, op Organization
 	return c.CanOrganization(ctx, orgID, op) || c.CanProject(ctx, projectID, pp)
 }
 
-func (c *authTokenClaims) composeOrgPermissions(ctx context.Context, orgID string) (*database.OrganizationRole, error) {
-	composite := &database.OrganizationRole{}
+func (c *authTokenClaims) OrganizationPermissions(ctx context.Context, orgID string) (*adminv1.OrganizationPermissions, error) {
+	c.orgCacheLock.Lock()
+	if perm, ok := c.orgPermissionsCache[orgID]; ok {
+		return perm, nil
+	}
+	c.orgCacheLock.Unlock()
+
+	composite := &adminv1.OrganizationPermissions{}
 	roles, err := c.admin.DB.ResolveOrganizationMemberUserRoles(ctx, c.token.OwnerID(), orgID)
 	if err != nil {
 		return nil, err
 	}
 	for _, role := range roles {
 		composite = unionOrgRoles(composite, role)
 	}
+
+	c.orgCacheLock.Lock()
+	c.orgPermissionsCache[orgID] = composite
+	c.orgCacheLock.Unlock()
 	return composite, nil
```

Comment:
It should take a lock for the entire duration of the func. This would still cause it to resolve user roles multiple times (reducing value of cache)